logserver/src/main/java/alluxio/logserver/AlluxioLog4jSocketNode.java

```diff
@@ -0,0 +1,148 @@
+              event.getMDC(AlluxioRemoteLogFilter.REMOTE_LOG_MDC_PROCESS_TYPE_KEY).toString());
+        }
+        Logger remoteLogger = hierarchy.getLogger(event.getLoggerName());
+        remoteLogger.callAppenders(event);
+      }
+    } catch (IOException e) {
+      // Something went wrong, cannot recover.
+      throw new RuntimeException(e);
+    } finally {
+      try {
+        mSocket.close();
+      } catch (Exception e) {
+        // Log the exception caused by closing socket.
+        LOG.warn("Failed to close client socket.");
+      }
+    }
+  }
+
+  /**
+   * Configure a {@link Hierarchy} instance used to retrive logger by name and maintain the logger
+   * hierarchy. {@link AlluxioLog4jSocketNode} instance can retrieve the logger to log incoming
+   * {@link org.apache.log4j.spi.LoggingEvent}s.
+   *
+   * @param processType type of the process sending this {@link LoggingEvent}
+   * @return a {@link Hierarchy} instance to retrieve logger
+   * @throws IOException if fails to create an {@link FileInputStream} to read log4j.properties
+   */
+  private LoggerRepository configureHierarchy(String processType)
+      throws IOException {
+    String inetAddressStr = mSocket.getInetAddress().getHostAddress();
+    Properties properties = new Properties();
+    File configFile;
+    try {
+      configFile = new File(new URI(System.getProperty("log4j.configuration")));
+    } catch (URISyntaxException e) {
+      // Alluxio log server cannot derive a valid path to log4j.properties. Since this
+      // properties file is global, we should throw an exception.
+      throw new RuntimeException("Cannot derive a valid URI to log4j.properties file.", e);
+    }
+    try (FileInputStream inputStream = new FileInputStream(configFile)) {
+      properties.load(inputStream);
+    }
+    // Assign Level.TRACE to level so that log server prints whatever log messages received from
+    // log clients. If the log server receives a LoggingEvent, it assumes that the remote
+    // log client has the intention of writing this LoggingEvent to logs. It does not make
+    // much sense for the log client to send the message over the network and wants the
+    // messsage to be discarded by the log server.
+    // With this assumption, since TRACE is the lowest logging level, we do not have to compare
+    // the level of the event and the level of the logger when the log server receives
+    // LoggingEvents.
+    Level level = Level.TRACE;
+    Hierarchy clientHierarchy = new Hierarchy(new RootLogger(level));
+    // Startup script should guarantee that mBaseLogsDir already exists.
+    String logDirectoryPath =
+        PathUtils.concatPath(mBaseLogsDir, processType.toLowerCase());
+    File logDirectory = new File(logDirectoryPath);
+    if (!logDirectory.exists()) {
+      logDirectory.mkdir();
+    }
+    String logFilePath = PathUtils.concatPath(logDirectoryPath, inetAddressStr + ".log");
+    properties.setProperty("log4j.rootLogger",
```

Comment:
nit: hardcoding the property key here looks a bit brittle to me... maybe make this key name as a constant? This is minor, up to you whether you want to address or not.

Reply:
Done.